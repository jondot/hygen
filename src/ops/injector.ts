import type { RenderedAction } from '../types'
import newline from '../newline'

const EOLRegex = /\r?\n/

const getPragmaticIndex = (pattern, lines, isBefore) => {
  const oneLineMatchIndex = lines.findIndex((l) => l.match(pattern))

  // joins the text and looks for line number,
  // we dont care about platform line-endings correctness other than joining/splitting
  // for all platforms
  if (oneLineMatchIndex < 0) {
    const fullText = lines.join('\n')
    const fullMatch = fullText.match(new RegExp(pattern, 'm'))

    if (fullMatch && fullMatch.length) {
      if (isBefore) {
        const fullTextUntilMatchStart = fullText.substring(0, fullMatch.index)
        return fullTextUntilMatchStart.split(EOLRegex).length - 1
      }
      const matchEndIndex = fullMatch.index + fullMatch.toString().length
      const fullTextUntilMatchEnd = fullText.substring(0, matchEndIndex)
      return fullTextUntilMatchEnd.split(EOLRegex).length
    }
  }

  return oneLineMatchIndex + (isBefore ? 0 : 1)
}
const locations = {
  at_line: (_) => _,
  prepend: (_) => 0,
  append: (_, lines) => lines.length - 1,
  before: (_, lines) => getPragmaticIndex(_, lines, true),
  after: (_, lines) => getPragmaticIndex(_, lines, false),
}
const indexesByLocation = (attributes: any, lines: string[]): number[] => {
  const pairs = Object.entries(attributes).filter(([k, _]) => locations[k]);
  if (pairs.length) {
    const indexes: number[] = []
    pairs.forEach((pair, i) => {
        const [k, v] = pair;
        indexes[i] = locations[k](v, lines);
    })
    return indexes;
  }
};
const injector = (action: RenderedAction, content: string): string => {
  const {
    attributes: { skip_if, eof_last },
    attributes,
    body,
  } = action
  const shouldSkip = skip_if && !!content.match(skip_if)

  if (!shouldSkip) {
    //
    // we care about producing platform-correct line endings.
    // however the "correct" line endings should be detected from the actual
    // CONTENT given, and not the underlying operating system.
    // this is similar to how a text editor behaves.
    //
    const NL = newline(content)
    const lines = content.split(NL)

    // returns -1 (end) if no attrs
    const idxs = indexesByLocation(attributes, lines)
    let idx = -1
    let deleteCount = 0
    if (idxs.length > 2) {
      throw new Error(`Too many injection attributes match '${attributes.join(',')}'! Don't know what to do.`)
    }
    if (idxs.length == 2) {
      idxs.sort((a, b) => a - b) // numeric ascending
      idx = idxs[0]
      deleteCount = idxs[1] - idxs[0]
    }
    if (idxs.length == 1) {
      idx = idxs[0]
      deleteCount = 0
    }

    const trimEOF = idx >= 0 && eof_last === false && /\r?\n$/.test(body)
    const insertEOF = idx >= 0 && eof_last === true && !/\r?\n$/.test(body)

    if (trimEOF) {
      lines.splice(idx, deleteCount, body.replace(/\r?\n$/, ''))
    } else if (insertEOF) {
      lines.splice(idx, deleteCount, `${body}${NL}`)
    } else if (idx >= 0) {
      lines.splice(idx, deleteCount, body)
    }
    return lines.join(NL)
  } else {
    return content
  }
}

export default injector
