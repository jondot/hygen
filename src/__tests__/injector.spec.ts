import injector from '../ops/injector'

const gemfile = `
    source 'http://rubygems.org'
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    `
describe('injector', () => {
  it('before rails', () => {
    expect(
      injector(
        {
          attributes: {
            before: "gem 'rails'",
          },
          body: "    gem 'kamikaze' # added by hygen",
        },
        gemfile,
      ),
    ).toMatchSnapshot()
  })
  it('after rails', () => {
    expect(
      injector(
        {
          attributes: {
            after: "gem 'rails'",
          },
          body: "    gem 'kamikaze' # added by hygen",
        },
        gemfile,
      ),
    ).toMatchSnapshot()
  })
  it('between rails', () => {
    expect(
      injector(
        {
          attributes: {
            after: "gem 'rails'",
            before: "gem 'httparty'",
          },
          body: "    gem 'kamikaze' # added by hygen",
        },
        gemfile,
      ),
    ).toMatchSnapshot()
  })
  it('prepend top of file', () => {
    expect(
      injector(
        {
          attributes: {
            prepend: true,
          },
          body: "    gem 'kamikaze' # added by hygen",
        },
        gemfile,
      ),
    ).toMatchSnapshot()
  })
  it('append bottom of file', () => {
    expect(
      injector(
        {
          attributes: {
            append: true,
          },
          body: "    gem 'kamikaze' # added by hygen",
        },
        gemfile,
      ),
    ).toMatchSnapshot()
  })
  it('at_index 2 (below "source")', () => {
    expect(
      injector(
        {
          attributes: {
            at_line: 2,
          },
          body: "    gem 'kamikaze' # added by hygen",
        },
        gemfile,
      ),
    ).toMatchSnapshot()
  })
  it('skip_if "source" exists', () => {
    expect(
      injector(
        {
          attributes: {
            skip_if: 'source',
            after: "gem 'rails'",
          },
          body: "    gem 'kamikaze' # added by hygen",
        },
        gemfile,
      ),
    ).toMatchSnapshot()
  })
  it('if eof_last is false remove empty line from the end of injection body', () => {
    expect(
      injector(
        {
          attributes: {
            after: "gem 'rails'",
            eof_last: false,
          },
          body: "    gem 'kamikaze' # added by hygen\n",
        },
        gemfile,
      ),
    ).toMatchSnapshot()
  })
  it('if eof_last is true insert empty line to injection body', () => {
    expect(
      injector(
        {
          attributes: {
            after: "gem 'rails'",
            eof_last: true,
          },
          body: "    gem 'kamikaze' # added by hygen",
        },
        gemfile,
      ),
    ).toMatchSnapshot()
  })
  it('correctly interpret multi-line after regex', () => {
    expect(
      injector(
        {
          attributes: {
            after: "rails[a-z\\:\\/\\.'\\s]*giri",
            eof_last: false,
          },
          body: "    gem 'kamikaze' # added by hygen",
        },
        gemfile,
      ),
    ).toMatchSnapshot()
  })
  it('correctly interpret multi-line before regex', () => {
    expect(
      injector(
        {
          attributes: {
            before: "rails[a-z\\:\\/\\.'\\s]*giri",
            eof_last: false,
          },
          body: "    gem 'kamikaze' # added by hygen",
        },
        gemfile,
      ),
    ).toMatchSnapshot()
  })
  it('correctly interpret multi-line skip_if regex', () => {
    expect(
      injector(
        {
          attributes: {
            skip_if: "rails[a-z\\:\\/\\.'\\s]*giri",
            after: "gem 'rails'",
          },
          body: "    gem 'kamikaze' # added by hygen",
        },
        gemfile,
      ),
    ).toMatchSnapshot()
  })
})
