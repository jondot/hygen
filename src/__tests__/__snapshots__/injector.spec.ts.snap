// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`injector after rails 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'kamikaze' # added by hygen
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injector append bottom of file 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    gem 'kamikaze' # added by hygen
    "
`;

exports[`injector at_index 2 (below "source") 1`] = `
"
    source 'http://rubygems.org'
    gem 'kamikaze' # added by hygen
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injector before rails 1`] = `
"
    source 'http://rubygems.org'
    gem 'kamikaze' # added by hygen
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injector between rails 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'kamikaze' # added by hygen
    gem 'httparty'

    "
`;

exports[`injector correctly interpret multi-line after regex 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'nokogiri'
    gem 'kamikaze' # added by hygen
    gem 'httparty'

    "
`;

exports[`injector correctly interpret multi-line before regex 1`] = `
"
    source 'http://rubygems.org'
    gem 'kamikaze' # added by hygen
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injector correctly interpret multi-line skip_if regex 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injector if eof_last is false remove empty line from the end of injection body 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'kamikaze' # added by hygen
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injector if eof_last is true insert empty line to injection body 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'kamikaze' # added by hygen

    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injector prepend top of file 1`] = `
"    gem 'kamikaze' # added by hygen

    source 'http://rubygems.org'
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injector skip_if "source" exists 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    "
`;
